Add a "Set as Featured Image" Button to Image Block

SiobhyB opened the pull request:
Partial fix for: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1011

`gutenberg-mobile`: wordpress-mobile/gutenberg-mobile#3116
`gutenberg`: https://github.com/WordPress/gutenberg/pull/31705

## Description

This PR adds a **Set as Featured** button to the image block's settings, with the purpose being to make it simpler for users to set a featured image within the post's editor. Users will also be able to **Remove as Featured** directly from the block's settings. 

A couple of notable changes to the native code have been highlighted below, with further details around the JS side highlighted in [the Gutenberg PR](https://github.com/WordPress/gutenberg/pull/31705). That PR can also be referred to as the central place for up-to-date screenshots and testing flows.

### OnSetFeaturedImageListener

[OnSetFeaturedImageListener](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-12a344ad8965875cb092fcc0c32a77e4909d9e5cb7092fbaf0d9ddc7d6c71d56R344
) listens for requests to set a featured image via an image block that are passed over the bridge. Please refer to [the Gutenberg PR ](https://github.com/WordPress/gutenberg/pull/31705)for more details of how this implemented from the JS side.

When a request is received from the native side, `onSetFeaturedImageButtonClicked()` [checks for whether a featured image has already been set](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-12a344ad8965875cb092fcc0c32a77e4909d9e5cb7092fbaf0d9ddc7d6c71d56R346), in which case `showFeaturedImageConfirmationDialog()` is called and users are able to confirm their wish to replace the featured image or not. If there isn't already an existing featured image or the user is removing an image as featured, then [the `setFeaturedImage()` function](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-12a344ad8965875cb092fcc0c32a77e4909d9e5cb7092fbaf0d9ddc7d6c71d56R809) is called directly.

_**Note: The code for setting a featured image has been iterated on as part of https://github.com/wordpress-mobile/WordPress-Android/pull/14521, with tracks events being added as part of that PR also. It'll be merged separately as an enhancement to this one.**_

### GutenbergDialogFragment

A [new `GutenbergDialogFragment` class](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-8889a43bfa80f8522889ae2f69695ac5eda5b29189bb29c9915a9c4825dae6e7) has been created as part of this PR. This class is based on [the `BasicFragmentDialog` class](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/posts/BasicFragmentDialog.kt), with some key differences being the removal of the third `setNeutralButton`, [an addition of a mediaId](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-8889a43bfa80f8522889ae2f69695ac5eda5b29189bb29c9915a9c4825dae6e7R39) to the constructor, and the ability for [the parent fragment to implement the positive/negative button interfaces](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-8889a43bfa80f8522889ae2f69695ac5eda5b29189bb29c9915a9c4825dae6e7R86). 

`GutenbergDialogFragment` [is called from the `showFeaturedImageConfirmationDialog()` function](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-12a344ad8965875cb092fcc0c32a77e4909d9e5cb7092fbaf0d9ddc7d6c71d56R795), to ensure that the dialog survives screen rotation. 

The **mediaId** associated with the image is [passed down from the `onSetFeaturedImageButtonClicked()` function](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-12a344ad8965875cb092fcc0c32a77e4909d9e5cb7092fbaf0d9ddc7d6c71d56R346) (which in turn retrieves it from the JS side).  

The functionality for when a user chooses to **replace** a featured image is defined in [the `onGutenbergDialogPositiveClicked()` function](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-12a344ad8965875cb092fcc0c32a77e4909d9e5cb7092fbaf0d9ddc7d6c71d56R1384) and the dialog is dismissed when the **cancel** button is clicked, as per [the `onGutenbergDialogNegativeClicked` function](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-12a344ad8965875cb092fcc0c32a77e4909d9e5cb7092fbaf0d9ddc7d6c71d56R1393).

As outlined in https://github.com/wordpress-mobile/WordPress-Android/issues/14676, there are other dialogs in the GutenbergEditorFragment file that are lost/dismissed with screen rotation. Part of the goal behind the new `GutenbergDialogFragment` class is to allow for easier refactoring of those other Dialogs, beginning with `showCancelMediaUploadDialog` and `showRetryMediaUploadDialog`, as those functions also accept a medaId.

## Testing

Please refer [to the Gutenberg PR](https://github.com/WordPress/gutenberg/pull/31705) as the central place for steps to tests and screenshots. Of note for the Android side, testing whether the **replace** dialog works as expected when the screen is rotated will be helpful.

## Regression Notes
1. Potential unintended areas of impact

The main unintended issues would be around the incorrect featured image being set/removed or other unintended/buggy issues with featured images.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

The flows outlined in [the Gutenberg PR](https://github.com/WordPress/gutenberg/pull/31705) were tested.

3. What automated tests I added (or what prevented me from doing so)

N/A

<hr />

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

SiobhyB committed:
Update strings.xml with text to be used

SiobhyB committed:
Add OnSetFeaturedImageListener()

This commit adds an OnSetFeaturedImageListener() in order to fetch a request to set a featured image from the JS side.

SiobhyB committed:
Add onSetFeaturedImageButtonClicked() function

onSetFeaturedImageButtonClicked() will be called when the option to set a featured image is clicked. It'll call showFeaturedImageConfirmationDialog(), which will be fleshed out in the next commit.

SiobhyB committed:
Add updateFeaturedImage() function

This commits adds "updateFeaturedImage" as an EditorFragmentListener method in EditPostActivity.java. The function enables an update to a post's featured image. By adding the function to "EditorFragmentAbstract.java", this can then be called from "GutenbergEditorFragment.java".

SiobhyB committed:
Add setFeaturedImage() and removeFeaturedImage() functions

These functions serve to either set a new featured image or remove an existing featured image, using the updateFeaturedImage() function defined in the previous commit. The ID of the new featured image (zero if removed) is then sent back to the JS side of the app via sendToJSFeaturedImageId().

SiobhyB committed:
Create dialog in showFeaturedImageConfirmationDialog()

The logic in showFeaturedImageConfirmationDialog() is fleshed out in this commit, with removeFeaturedImage() being called if the featured image is removed, setFeaturedImage() being called by itself if a new featured image is set, and an AlertDialog if there is a replacement.

SiobhyB committed:
Send a notice depending on whether a featured image is set or removed

The showNotice() function is utilised to send a notice to the user depending on whether a featured image is set or removed.

SiobhyB committed:
Update reference to latest Gutenberg Mobile commit

jd-alexander committed:
added an update featured image use case.

jd-alexander committed:
Created mechanisms to track when the gb featured image is picked

jd-alexander committed:
Integrated the new use case and approach to EditPost and Post Settings

SiobhyB committed:
Update remaining references to "IMAGE_PICKED" to "IMAGE_PICKED_POST_SETTINGS"

SiobhyB committed:
Remove unused imports from EditPostSettingsFragment.java

SiobhyB committed:
Merge branch 'develop' into gb/1011-set-featured-button

SiobhyB committed:
Remove removeFeaturedImage() function and consolidate into a single setFeaturedImage() function

SiobhyB committed:
Bump Gutenberg Mobile

SiobhyB committed:
Merge branch 'develop' into gb/1011-set-featured-button

SiobhyB committed:
Refactor logic for around showFeaturedImageConfirmationDialog()

With this commit, the showFeaturedImageConfirmationDialog() is refactored so that it only includes code for building the confirmation dialog.

The if/else logic for determining whether the dialog displays is moved to the onSetFeaturedImageButtonClicked() function.

The goal is to make the purpose of each function clearer.

SiobhyB committed:
Update build.gradle with latest Gutenberg Mobile reference

SiobhyB committed:
Update build.gradle to reference latest commit from Gutenberg Mobile

SiobhyB committed:
Merge branch 'develop' into gb/1011-set-featured-button

SiobhyB committed:
Merge branch 'develop' into gb/1011-set-featured-button

SiobhyB committed:
Merge branch 'develop' into gb/1011-set-featured-button

SiobhyB committed:
Merge branch 'develop' into gb/1011-set-featured-button

SiobhyB committed:
Merge branch 'develop' into gb/1011-set-featured-button

SiobhyB committed:
Update reference to latest Gutenberg Mobile commit

SiobhyB committed:
Merge branch 'develop' into gb/1011-set-featured-button

SiobhyB committed:
Merge branch 'gb/1011-set-featured-button' of github.com:wordpress-mobile/WordPress-Android into gb/1011-set-featured-button

SiobhyB committed:
Update reference to latest Gutenberg Mobile commit

SiobhyB committed:
Bump Gutenberg Mobile reference

jd-alexander committed:
Updated gutenberg-mobile ref.

jd-alexander committed:
added inject annotation to the constructor of the use case.

jd-alexander committed:
Updated gutenberg-mobile ref.

jd-alexander committed:
removed comment about disabling Audio block since it is now on all sites

jd-alexander committed:
Merge branch 'gb/1011-set-featured-button' into gb/1011-set-featured-button-refactor

jd-alexander committed:
formatted the use case.

jd-alexander committed:
sync stories submodule.

SiobhyB committed:
Create GutenbergDialogFragment.kt using BasicDialogFragment.kt as a template

The code for this DialogFragment is copied from the BasicDialogFragment file, with the neutral button omitted: https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/posts/BasicFragmentDialog.kt

The code will continue to be edited for use with dialogs in GutenbergEditorFragment.

SiobhyB committed:
Add mediaId to GutenbergDialogFragment's constructor

SiobhyB committed:
Edit "onNegativeClicked" and "onPositiveClicked" so that interface can be used in parent fragment, rather than host activity

SiobhyB committed:
Implement BasicDialogPositiveClickInterface and BasicDialogNegativeClickInterface in GutenbergEditorFragment.java

SiobhyB committed:
Refactor showFeaturedImageConfirmationDialog() to use new GutenbergDialogFragment

SiobhyB committed:
Add "org.jetbrains.annotations.NotNull" import

This is needed in order for the @NotNull annotations in onPositiveClicked() and onNegativeClicked() to work without error.

SiobhyB committed:
Fill out "onPositiveClicked" function

When the positive button in the "Replace Featured" dialog is clicked, setFeaturedImage() should be called.

SiobhyB committed:
Fill out onNegativeClicked() function

When the negative button in the "Replace Featured" dialog is clicked, the dialog should just simply be dismissed.

SiobhyB committed:
Rename instances of "BasicDialog" to "GutenbergDialog"

SiobhyB committed:
Update reference to Gutenberg Mobile

SiobhyB committed:
Rename "onPositiveClicked" and "onNegativeClicked"

These functions have been renamed to "onGutenbergDialogPositiveClicked" and "onGutenbergDialogNegativeClicked" respectively, in order to make their purpose clearer.

SiobhyB commented:
Status update: I'm currently investigating an issue where `mFeaturedImageId` is reset to zero when following these steps:

1. Open a new post that already has a featured image, or set a featured image on the post if it doesn't already have one.
2. Select a different image block, tap the cog/gear icon, and then **Set as Featured** image.
3. When the **replace** dialog pops up, switch the screen's orientation before tapping the **replace** option.
4. Confirm the new image has been set as featured by navigating to _Post Settings_ and observing the **Featured** banner that overlays it within the editor.
5. Switch the orientation of the screen again. Select a different image block, tap the cog/gear icon, and then **Set as Featured** image.
6. Notice how the new image is automatically set as featured, with no _replace_ dialog displaying.

Even though the correct images are being set as featured in the above flow, the replace dialog should still display at step number 6. _Log.d_ shows that `mFeaturedImageId` is being reset to zero in this scenario, which explains why the dialog isn't popping up. The images themselves are being set as featured, so this doesn't seem like a major bug and more of an edge case, but would still be nice for the behaviour to be consistent. I'll keep investigating. 🕵️‍♀️

SiobhyB committed:
Merge branch 'develop' into gb/1011-set-featured-button

hypest commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergDialogFragment.kt:
```diff
@@ -0,0 +1,129 @@
+import com.google.android.material.dialog.MaterialAlertDialogBuilder
+
+class GutenbergDialogFragment() : AppCompatDialogFragment() {
+    private lateinit var mTag: String
```
👋 @SiobhyB , only took a quick glance here, and since I know you've taken inspiration from the BasicDialog implementation I wonder if you can elaborate on the choice of having these dialog "properties" (Tag, message, etc) be members of the Fragment while [they are part of a model class over at the BasicDialog implementation](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/posts/BasicDialogViewModel.kt#L38-L45). Not saying there's an issue with this, haven't really checked deep enough yet, just wondering what's the context/insight on this. Thanks!

SiobhyB replied:
Ah, I used [BasicFragmentDialog.kt](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/posts/BasicFragmentDialog.kt) as the template, which implements the "properties" in this way. I've gone ahead to edit the commit message that incorrectly referenced BasicDialog.kt rather than BasicFragmentDialog.kt, sorry for the confusion!

hypest replied:
Ahh, I see, thanks for the update!

jd-alexander replied:
@SiobhyB I have taken a look at the `BasicFragmentDialog.kt` you mentioned as your point of reference and compared it to the `GutenbergDialogFragment` and the main difference I saw was the inclusion of the `mediaId`. I am thinking we could reuse `BasicFragmentDialog.kt` and track the `mediaId` within the `GutenbergDialogFragment` itself. Wdyt?

SiobhyB replied:
@jd-alexander, the new class also removes the third `setNeutralButton` and adds [the ability for the parent fragment to implement the positive/negative button interfaces](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-8889a43bfa80f8522889ae2f69695ac5eda5b29189bb29c9915a9c4825dae6e7R86) (rather than [the host activity](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/posts/BasicFragmentDialog.kt#L83), which is the default with `BasicFragmentDialog.kt` and would require changes to `EditPostActivity` in our case). Would you still lean towards re-using `BasicFragmentDialog.kt` over a new class with those additional changes considered?

jd-alexander replied:
Ah, thanks for expounding on that. We can go ahead and keep this class then. The only thing I would change is probably the `mediaId` references since the name of the class is `GutenbergDialogFragment ` and there could be other use cases for it in the future. In the interest of time and not overcomplicating the issue we could even remove the `media` and make it just `id`. `STATE_KEY_MEDIA_ID = "state_key_media_id"` ==> `STATE_KEY_ID = "state_key_id"` Ideally, I would utilize generics, so we could make it a `param` value that isn't bound to a type and in this case, we could pass the `Int` but yes, we don't have to do that optimization at this time so we could probably evaluate changing the name for now. Let me know what you think.

SiobhyB replied:
> In the interest of time and not overcomplicating the issue we could even remove the `media` and make it just `id`.

That makes sense! I've gone ahead to implement that change in https://github.com/wordpress-mobile/WordPress-Android/pull/14503/commits/40310c385d7c435912942c2a9b0b2ea6610d1445.

> Ideally, I would utilize generics, so we could make it a param value that isn't bound to a type and in this case, we could pass the Int but yes, we don't have to do that optimization at this time so we could probably evaluate changing the name for now. Let me know what you think.

That sounds great, I've created [an issue to track other places where this class](https://github.com/wordpress-mobile/WordPress-Android/issues/14676) can be used and have made a note of your comments for optimizing it in the comments here: https://github.com/wordpress-mobile/WordPress-Android/issues/14676#issuecomment-852024183. Does iterating on the class as part of that separate issue seem like a reasonable approach to you?

hypest replied:
Makes sense to iterate in a separate effort, outside this PR 👍

SiobhyB committed:
Reformat code in GutenbergDialogFragment.kt

SiobhyB committed:
Update build.grade with latest reference to Gutenberg Mobile

SiobhyB committed:
Save "mFeaturedImageId" in savedInstanceState() and onSaveInstanceState() methods

Saving "mFeaturedImageId" in these methods prevents its value being lost when the screen is rotated.

This addresses the bug outlined here: https://github.com/wordpress-mobile/WordPress-Android/pull/14503#issuecomment-844950869

SiobhyB commented:
[This bug](https://github.com/wordpress-mobile/WordPress-Android/pull/14503#issuecomment-844950869) has been fixed with the following commit: https://github.com/wordpress-mobile/WordPress-Android/pull/14503/commits/f57eeaf8703a2ef816cc98949a54b99ab6432584

I've asked for some help on Slack regarding the failing test: p1621524064033200-slack-C02QANACA

Edited to add: The issue causing the failing test has been resolved now. 🎉

SiobhyB committed:
Remove redundant comma

SiobhyB committed:
Add checks to ensure GutenbergDialogPositiveClickInterface and GutenbergDialogNegativeClickInterface (if not null) are implemented in the parent fragment

SiobhyB committed:
Don't allow "mPositiveButtonLabel" to be null

SiobhyB committed:
Add "lateinit" modifier to mPositiveButtonLabel

SiobhyB committed:
Merge branch 'gb/1011-set-featured-button' into gb/1011-set-featured-button-refactor

SiobhyB committed:
Add Tracks event that fires when a featured image is removed from the Gutenberg editor

This commit adds a "IMAGE_REMOVED_GUTENBERG_EDITOR" event to complement the "IMAGE_PICKED_GUTENBERG_EDITOR" event, for when an image is removed via the image block.

hypest commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergDialogFragment.kt:
```diff
@@ -0,0 +1,139 @@
+        mPositiveButtonLabel?.let {
+            builder.setPositiveButton(mPositiveButtonLabel) { _, _ ->
+                dismissedByPositiveButton = true
+                val parentFragment: Fragment? = parentFragment
```
Hey Siobhan, I know this pattern of having a same-name local const picking up the value from a getter (`getParentFragment()` in this case) [is used in similar places](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/posts/BasicFragmentDialog.kt#L83-L86), but makes me a bit uncomfortable when reading the code. Trusted, in AndroidStudio with its syntax highlighting it's kinda easy to see the difference between the two, but it's not super intuitive otherwise.

I was thinking, would you be interested in using the following to do the same null-check and casting directly, without the temporary val?

```kotlin
               (parentFragment as? GutenbergDialogPositiveClickInterface)?.let {
                    it.onGutenbergDialogPositiveClicked(mTag, mMediaId);
                }
```

SiobhyB replied:
Thank you for this, it led me down a road of learning [a bit more about Kotlin](https://discuss.kotlinlang.org/t/let-vs-if-not-null/3542/11) and I've gone ahead to implement your suggestion. :)

hypest replied:
Ah, nice! While at it, wdyt about doing the same treatment to [this null-check and cast case in `onDismiss()`](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-8889a43bfa80f8522889ae2f69695ac5eda5b29189bb29c9915a9c4825dae6e7R116-R122) too?

SiobhyB replied:
Ah, good point! I've gone ahead to do that in https://github.com/wordpress-mobile/WordPress-Android/pull/14503/commits/7924341c7b430ec48d6fbb56d2f4d6a8866491c9, as part of that commit I also changed the reference to `activity` to `parentFragment`, to match the way the other interfaces are implemented. Let me know if you spot any issues with that!

hypest replied:
Thanks! I [left a comment on the commit](https://github.com/wordpress-mobile/WordPress-Android/commit/7924341c7b430ec48d6fbb56d2f4d6a8866491c9#r51332384).

hypest commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:
```diff
@@ -333,6 +343,16 @@ public void onQueryCurrentProgressForUploadingMedia() {
+                        } else if (mFeaturedImageId != mediaId) {
+                            showFeaturedImageConfirmationDialog(mediaId);
+                        }
+                    }
```
Looks like the function only implicitly covers the case where the IDs are non-zero but equal. I know that that case is practically a no-op case (we don't want anything to happen if we're trying to set the same image as the current featured one) but it's best if we leave a comment for that to make it clear. Something like `// don't do anything if mediaId is the same as the current featuredImageId`.

Alternatively, and a bit of a nit, what do you think about restructuring the code in such a way that it's more clear what's going on with the special values (`0`) and the various combinations, like this:

```kotlin
public void onSetFeaturedImageButtonClicked(int mediaId) {
    if (mediaId == mFeaturedImageId) {
        // nothing special to do, trying to set the image that's already set as featured
        return;
    }

    if (mediaId == 0) {
        // user tries to clear the featured image setting
        setFeaturedImage(mediaId);
        return;
    }

    if (mFeaturedImageId == 0) {
        // current featured image is not set so, go ahead and set it to the provided one
        setFeaturedImage(mediaId);
        return;
    }

    // ask the user to confirm changing the featured image since there's already one set
    showFeaturedImageConfirmationDialog(mediaId);
}
```

It's of course more verbose, but that's actually part of the goal, to use verbosity for the sake of readability and simplifying understanding the combinations.

SiobhyB replied:
Implemented!

> It's of course more verbose, but that's actually part of the goal, to use verbosity for the sake of readability and simplifying understanding the combinations.

This is useful insight, thank you! I think knowing when to favour verbosity is something I'm going to need to work at.

hypest replied:
> I think knowing when to favour verbosity is something I'm going to need to work at.

Thanks for looking at this that way! And you are 100% correct with "when to favour" as it's always a balance or compromise between too verbose and too compacted without it being clear usually which side one should lean to 😬

hypest commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:
```diff
@@ -774,6 +794,33 @@ public void onClick(DialogInterface dialog, int id) {
+        mEditorFragmentListener.updateFeaturedImage(mediaId, false);
+        // Necessary to update image in Post Settings
+        setFeaturedImageId(mediaId);
+        sendToJSFeaturedImageId(mediaId);
```
Hmm, do we need the explicit `sendToJSFeaturedImageId()` call here? Not sure if I'm missing some codepath but looks like the `updateFeaturedImage()` two lines up is including a call to send the data to the JS side? 🤔

SiobhyB replied:
Aha, yes, I missed that. 🤦‍♀️ Thank you for catching! I've removed it in the latest commit.

hypest replied:
> I've removed it in the latest commit

May I offer a suggestion (preference?) for how to link and close "the loop"? I personally add messages like "Fixed with <this commit hash link>" ([example here](https://github.com/wordpress-mobile/WordPress-Android/pull/6196#discussion_r127898836)) to save everyone a couple of seconds from scrolling to find the commit in question. I actually do that in all threads that end up in code changes to help focus the PR review. Of course, the reviewer will probably have a look at the overall diff as well but often times it helps to quickly check the commit that addresses a particular feedback and move on, until another overall pass.

SiobhyB replied:
Noted for the future! Thank you so much, I appreciate tips around etiquette like that and see how that small change would make life a lot easier for the review. 🙇‍♀️

hypest commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:
```diff
@@ -774,6 +794,33 @@ public void onClick(DialogInterface dialog, int id) {
+    private void setFeaturedImage(int mediaId) {
+        mEditorFragmentListener.updateFeaturedImage(mediaId, false);
+        // Necessary to update image in Post Settings
+        setFeaturedImageId(mediaId);
```
👋 Siobhan! Assuming the comment (`// Necessary to update image in the Post Settings`) is about the `setFeaturedImageId()` call, can you elaborate on the necessasity? I'm trying to follow the code and looks to me that [the member variable](https://github.com/wordpress-mobile/WordPress-Android/blob/74323e97c25ac24447dce5030f01f84f900eaf6b/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/EditorFragmentAbstract.java#L143-L145) doesn't play a huge role and doesn't seem to be involved [when saving the data](https://github.com/wordpress-mobile/WordPress-Android/blob/332ca7f48d3142fae0267beecf0212166adc8fc6/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java#L999-L1020) 🤔. Maybe I'm missing a use?

jd-alexander replied:
Hi @hypest to add to the conversation here, I can see where setting member variable would be useful if it's being persisted across configuration changes. 

For the "Necessary to update image in the Post Settings" comment, this will be removed once https://github.com/wordpress-mobile/WordPress-Android/pull/14521 is merged into this one. It is now under review.

hypest replied:
> would be useful if it's being persisted across configuration changes.

Hmm, the comment would still need some rephrasing then since it currently mentions "Post Settings" which makes me think of persisting in the local device prefs or the DB, which is different than preserving it on configuration changes (rotations for example). In any case, I feel that either the code doesn't correspond to the comment anymore, or vice versa 🙂

SiobhyB replied:
It looks like the original bug referenced in the comment is no longer reproducible, but the functionality's still breaking if I comment out the call to `setFeaturedImageId`. I was able to reproduce the following with that line commented out:

* Toggle between **Set as Featured** and **Remove as Featured** on any image block, or between more than one image block.
* Notice how sometimes the button is unresponsive or the **replace** dialog pops up when there isn't actually a current featured image.

From logging, I see that `mFeaturedImageId` isn't always automatically updated when going through that flow. I'm not exactly sure if that's okay or expected. Is there another approach for updating that variable that should/could be taken here?

For now, I've updated the `Necessary to update image in Post Settings` to `Ensure "mFeaturedImageId" member variable is updated to specified "mediaId"`, in order to try to be clearer. Looking forward to hearing any extra thoughts!

SiobhyB replied:
@hypest, we must have been drafting our comments at the same time and I missed yours until after I clicked published. :) You're right that the comment is now outdated, interested to hear what you think of the updated phrasing.

hypest replied:
> `Ensure "mFeaturedImageId" member variable is updated to specified "mediaId"`

Hmm, I guess the comment now is more accurate, but unfortunately ended up not adding any context. I mean, the `setFeaturedImageId(mediaId)` itself is nothing more than a classic (Java) setter so it is in fact already conveying the message "let's set the `FeaturedImageId` to `mediaId`. With a line comment, we are seeking to add more info that is not obvious just by looking at the code. So in this case, I'd expect a comment that would answer the question: "**Why** are we setting `mFeaturedImageId` to `mediaId`?"

I understand that perhaps we are not quite sure yet, and I guess that's what we need to do, to make sure we understand the code and the need. Let's continue reading and researching the code to see what's going and. That should also help answer your other question:
> "From logging, I see that mFeaturedImageId isn't always automatically updated when going through that flow. I'm not exactly sure if that's okay or expected. Is there another approach for updating that variable that should/could be taken here?".

SiobhyB replied:
>  With a line comment, we are seeking to add more info that is not obvious just by looking at the code. So in this case, I'd expect a comment that would answer the question: "Why are we setting mFeaturedImageId to mediaId?"

@hypest, from my side, I think I'm missing something. It feels correct to me that `setFeaturedImageId` would be called here, as I don't see anywhere else in the file where `mFeaturedImageId` is being updated. I wouldn't expect it to be automatically updated without a function call. Would you be able to help me understand why this doesn't look like you'd expect? That may help me to understand the sort of context that needs to be added or what to explore further in the code.

hypest replied:
Ah, if there's no special meaning around the appearance of `setFeaturedImageId()` here (now that we've updated the original comment that was trying to elevate the importance) the we better just go ahead and remove the line comment as it doesn't add any context.

In other words, I'm not trying to say that the call should be somewhere else or not happening at all but, wondering if there's something special about it. If yes it makes sense to capture it in the line comment, otherwise it's actually better to remove the comment. Does that help?

SiobhyB replied:
Got it, thank you for the extra explanation! This thread's definitely helped me to see how comments can cause confusion and this one is pretty unecessary given the function itself is fairly self-explanatory. I've gone ahead to remove it in https://github.com/wordpress-mobile/WordPress-Android/pull/14503/commits/36b6e2ee848e98baab166230f2e1de04366b92dd.

hypest commented on libs/editor/WordPressEditor/src/main/res/values/strings.xml:
```diff
@@ -154,6 +154,13 @@
 
-
+    <!-- Featured image replacement -->
+    <string name="replace_current_title">Replace current?</string>
```
Wdyt about tweaking the title to not include an implied reference to the body (`current`)? I feel it might be tougher to translate. Some ideas: "Confirm replace?" or just "Replace" even?

While at it, maybe all the string names could be prefixed with `featured_image_` to provide context?

SiobhyB replied:
I agree that, upon reflection, the names here a bit confusing! I've made some changes so that they're now prefixed with `featured_image_replace_dialog_`, following the lead of some of the other strings for dialogs in the `strings.xml` file. Do you think these new names seem okay?

hypest replied:
> Do you think these new names seem okay?

Yeap, names look good to me!

Just making sure, do you have thoughts about rephrasing the string as well?

SiobhyB replied:
Ah, yes, I agree with the reasoning for rephrasing. I went with `Replace?` in https://github.com/wordpress-mobile/WordPress-Android/pull/14503/commits/b4dc8252d0706ec223a38614546cffbb94d5c292, as that seemed to flow best in my mind, let me know if you think anything else is needed here.

hypest replied:
Yeah, "Replace?" works for me too, thanks!

SiobhyB committed:
Refactor null-checks and casting in setPositiveButton() and setNegativeButton()

This commit refeactors, setPositiveButton() and setNegativeButton() so that "?.let" is used instead of  "!= null" as a null-check for the parentFragment.

SiobhyB committed:
Refactor if/else statement in onSetFeaturedImageButtonClicked()

With this commit, the if/else statement is refactored to a series of single if statements, with accompanying comments. The goal is to make the code more readable and for it to be clearer what the purpose of each combination is. It also adds an explicit if statement for any cases where the mediaId may equal mFeaturedImageId.

SiobhyB committed:
Remove call to sendToJSFeaturedImageId() from within setFeaturedImage()

sendToJSFeaturedImageId() doesn't need to be called here, as it's already called directly from updateFeaturedImage() function two lines up.

SiobhyB committed:
Remove unnecessary semi-colons

SiobhyB committed:
Refine string names

SiobhyB committed:
Merge branch 'develop' into gb/1011-set-featured-button

SiobhyB committed:
Update string name being passed to showNotice() calls

jd-alexander commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:
```diff
@@ -333,6 +343,30 @@ public void onQueryCurrentProgressForUploadingMedia() {
+                            return;
+                        }
+
+                        if (mediaId == 0) {
```
I am thinking we could hardcode this `0` which serves as the `DEFAULT_MEDIA_ID` so it's easier to identify its purpose and if we need to change its value it is in a reusable state.

hypest replied:
Good point about naming the special value. In this case, I guess it's probably closer to a `MEDIA_ID_NO_FEATURED_IMAGE_SET` that denotes the special interpretation of the value, right?

SiobhyB replied:
That makes a lot of sense! I've introduced that constant in https://github.com/wordpress-mobile/WordPress-Android/pull/14503/commits/9630c2955da873dbbd033849a6d1494876bbb582. I've gone ahead to use the `MEDIA_ID_NO_FEATURED_IMAGE_SET` suggestion for naming and used it in three different places, where there were checks for zero. How does that seem?

jd-alexander replied:
> That makes a lot of sense! I've introduced that constant in 9630c29. I've gone ahead to use the MEDIA_ID_NO_FEATURED_IMAGE_SET suggestion for naming and used it in three different places, where there were checks for zero. How does that seem?

Looking good @SiobhyB

hypest replied:
Nice! Any opportunities to introduce the const (or define in a common place and reuse) on the "sending" side as well? Like, in the places where functions _use_ the `0` to denote clearing the setting, no just when checking against it.

SiobhyB replied:
I've gone ahead to make some changes to the button on the JS side here https://github.com/WordPress/gutenberg/pull/31705#pullrequestreview-667361512

If there's a better way for defining the constant there, I'm all ears. :D

hypest replied:
Looks like all comments/ideasof this thread have been addressed so, I'll mark it as resolved :)

jd-alexander commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:
```diff
@@ -333,6 +343,30 @@ public void onQueryCurrentProgressForUploadingMedia() {
+
+                        if (mFeaturedImageId == 0) {
+                            // current featured image is not set so, go ahead and set it to the provided one
+                            setFeaturedImage(mediaId);
```
Within `EditorFragmentAbstract` `mFeaturedImageSupported` is being persisted via the `savedInstanceState` bundle. We may need to persist `mFeaturedImageId`. Have you tried rotating the device while a `mFeaturedImageId` is set, attempt to set the same media id, and see if the function returns? Let me know what that is like. I will do another test of this PR tomorrow.

SiobhyB replied:
@jd-alexander, `mFeaturedImageId` was added to `savedInstanceState` within `GutenbergEditorFragment` in https://github.com/wordpress-mobile/WordPress-Android/pull/14503/commits/f57eeaf8703a2ef816cc98949a54b99ab6432584, which I believe fixes any issues with it being reset with screen rotation. Would it be better practice to do this in `EditorFragmentAbstract` rather than `GutenbergEditorFragment`?

jd-alexander replied:
Ah, I didn't notice that. Thanks for pointing it out.

SiobhyB committed:
Update comment to clarify reasoning behind "setFeaturedImageId" call

SiobhyB committed:
Reword "featured_image_replace_dialog_title" string

SiobhyB committed:
Introduce "MEDIA_ID_NO_FEATURED_IMAGE_SET" const

With this commit, a "MEDIA_ID_NO_FEATURED_IMAGE_SET" constant is introduced and used in checks where a featured image is being removed or is already set to zero.

The purpose behind introducing a constant is to make its purpose clearer and make it easier to change in the future, if needed.

SiobhyB committed:
Refactor onDismiss() in GutenbergDialogFragment

The function is refactor so that its interface can used in parent fragment, rather than host activity. The original code is also refactored so that "?.let" is used instead of  "!= null" as a null-check for the parent fragment.

SiobhyB committed:
Bump reference to Gutenberg Mobile

SiobhyB committed:
Merge branch 'develop' into gb/1011-set-featured-button

SiobhyB committed:
Remove unecessary comment before 'setFeaturedImageId'

As the functionality of setFeaturedImageId() is fairly self-explanatory, the current comment before it doesn't add any value and has the potential to cause confusion around the function's importance.

SiobhyB committed:
Fix 'exceeded max line length' ktlint error

A line in the onDismiss() function exceeded the max line length allowed, so this commit splits that line up.

SiobhyB committed:
Remove conditional surrounding 'GutenbergDialogOnDismissByOutsideTouchInterface' in onDismiss()

The 'parentFragment is GutenbergDialogOnDismissByOutsideTouchInterface' was not serving a specific purpose that the refactored 'parentFragment as? GutenbergDialogOnDismissByOutsideTouchInterface' doesn't.

SiobhyB committed:
Update build.gradle reference to latest commit from Gutenberg Mobile

jd-alexander committed:
Merge branch 'develop' into gutenberg/audio-block-capability

jd-alexander committed:
Merge branch 'develop' into gutenberg/audio-block-capability

jd-alexander committed:
renamed capability

jd-alexander committed:
sync stories submodule hash.

jd-alexander committed:
updated gutenberg-mobile ref.

hypest reviewed (approved):
Awesome work here @SiobhyB , thanks!

This looks good to me and I will leave it to other reviewers to finish their rounds before we start merging 👍

SiobhyB committed:
Bump Gutenberg Mobile reference

SiobhyB committed:
Merge branch 'develop' into gb/1011-set-featured-button

fluiddot committed:
Update Gutenberg Mobile ref

jd-alexander committed:
reformatted use case

jd-alexander committed:
Merge pull request #14521 from wordpress-mobile/gb/1011-set-featured-button-refactor

Add a "Set as Featured Image" Button to Image Block - Refactor for Featured Image PostModel update

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -2484,12 +2485,32 @@ protected void setPostContentFromShareAction() {
             }
+
+            int postId = getEditPostRepository().getId();
+            if (mediaId == 0) {
```
@SiobhyB wdyt about importing `MEDIA_ID_NO_FEATURED_IMAGE_SET ` here?

SiobhyB replied:
@jd-alexander, that makes sense to me! I've gone ahead to import that constant in https://github.com/wordpress-mobile/WordPress-Android/pull/14503/commits/fb3ba9548da20bea53f50ca56e8bc6fc584e877e. Let me know if anything looks off to you. :)

hypest replied:
Looks good to me!

jd-alexander reviewed (approved):
Thanks for these changes @SiobhyB I verified that the functionality here works as expected by utilizing the test cases on the Gutenberg PR. I left a few comments but none of them were blocking and the changes are minor, so this can be approved for merge. 🚢 

The lint error is probably fixed on `develop` so you could merge it in here to see if that resolves it.

fluiddot committed:
Update Gutenberg Mobile ref

SiobhyB committed:
Merge branch 'develop' into gb/1011-set-featured-button

SiobhyB committed:
Import constant for use in EditPostActivity

With this commit, MEDIA_ID_NO_FEATURED_IMAGE_SET is imported to EditPostActivity, so that it can be used in a conditional checking for whether a featured image is being set.

SiobhyB committed:
Rename references to 'mediaId' to 'id'

With this commit, references to 'mediaId' are updated to 'id' in the GutenbergDialogFragment.kt class. The purpose is to make the name more generic so that the class is more re-usable.

SiobhyB commented:
@jd-alexander, thank you for this. 🙇‍♀️ I've gone ahead to implement your suggestions, let me know what you think and I'll go ahead with the merge waterfall if all looks okay to you. :)

hypest commented:
👋 @SiobhyB , Joel will be on AFK for the next few weeks so, I'll go ahead and take over the last steps to finalize the review here.

I gave the PR a try locally using local builds and looks like I run into an issue with multiple gradle plugin versions detected:
```
Build file '/Users/stefanos/proj/a8c/android/tmp/t2/WordPress-Android/gutenberg-mobile/gutenberg/node_modules/@react-native-community/masked-view/android/build.gradle' line: 17

* What went wrong:
A problem occurred evaluating project ':android:@react-native-community_masked-view'.
> Failed to apply plugin [id 'com.android.internal.library']
   > Using multiple versions of the Android Gradle plugin in the same build is not allowed
```

The binary (composite) build works OK, only the local source mode encounters the issue. Perhaps some change in WPAndroid or gutenberg-mobile has introduce the mismatch, can you have a look? I don't think this PR has anything to do with the mismatch, but we might as well avoid merging as-is so we don't introduce a development flow breakage. wdyt?

SiobhyB committed:
Reword dialog's title for clarity.

The title of the replace dialog is updated from "Replace?" to "Featured Image Already Set" with this commit. This brings it more inline with how titles are presented in other areas of the app (questions aren't generally used for titles). The overall goal is to improve clarity.

SiobhyB committed:
Undo title case in notices for consistency

Following feedback, it makes sense to undo title case that's applied to notices, as they're not strictly headings/titles. This will help with consistency in how notices are capitalised throughout the rest of the apps.

SiobhyB committed:
Update to latest Gutenberg Mobile reference

SiobhyB commented:
@hypest, thank you for the heads up!  

> Perhaps some change in WPAndroid or gutenberg-mobile has introduce the mismatch, can you have a look? I don't think this PR has anything to do with the mismatch, but we might as well avoid merging as-is so we don't introduce a development flow breakage. wdyt?

There was a little bit of a window between [Android](https://github.com/wordpress-mobile/WordPress-Android/pull/14695) and [Gutenberg](https://github.com/WordPress/gutenberg/pull/32144) updating to 4.0.2, which explains what happened here. I've pulled the latest changes from Gutenberg now, so all should work again. :)

hypest commented:
@SiobhyB , do you perhaps get this AndroidStudio issue `Cannot access 'androidx.activity.result.ActivityResultCaller' which is a supertype of 'org.wordpress.android.editor.gutenberg.GutenbergDialogFragment'. Check your module classpath for missing or conflicting dependencies`? See screenshot too. WPAndroid builds fine both via cli gradle and AndroidStudio (and dialog works fine too) so, not a blocker for the PR, just wondering if it's on my side alone or not.

![Screenshot 2021-06-03 at 11 27 50](https://user-images.githubusercontent.com/1032913/120614604-1600fc80-c460-11eb-9f8b-022ffc362a0e.png)

hypest reviewed (approved):
Let's get this merged @SiobhyB !

I think the [libraries misalignment error AndroidStudio is giving out](https://github.com/wordpress-mobile/WordPress-Android/pull/14503#issuecomment-853694243) is not a blocker but, can you please open a ticket and check it out in a separate PR? WDYT?

SiobhyB commented:
Thank you for this, @hypest! I'm going to begin the domino. :) 

Just noting our Slack conversation, also, for posterity: p1622711015069900/1622706783.069600-slack-C01NFTM4BMY

Although I'm not currently able to replicate the error you're seeing (likely due to an older version of AS), I'll look further into this and create a GH issue. ✅

fluiddot committed:
Merge branch 'develop' into gutenberg/audio-block-capability

# Conflicts:
#	build.gradle

fluiddot committed:
Update Gutenberg Mobile ref

fluiddot committed:
Update Gutenberg Mobile ref

SiobhyB committed:
Bump Gutenberg Mobile reference

fluiddot committed:
Merge branch 'gutenberg/audio-block-capability' into gb/1011-set-featured-button

# Conflicts:
#	build.gradle

fluiddot commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -2302,9 +2304,7 @@ private GutenbergPropsBuilder getGutenbergPropsBuilder() {
                 isUnsupportedBlockEditorEnabled,
                 unsupportedBlockEditorSwitch,
-                !isFreeWPCom, // Disable audio block until it's usable on free sites via "Insert from URL" capability
-                // Only enable reusable block in WP.com sites until the issue
-                // (https://github.com/wordpress-mobile/gutenberg-mobile/issues/3457) in self-hosted sites is fixed
+                !isFreeWPCom,
```
These changes come from this [PR](https://github.com/wordpress-mobile/WordPress-Android/pull/14680), that it was required to merge it here due to conflicts in the Gutenberg reference.

fluiddot commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergPropsBuilder.kt:
```diff
@@ -14,7 +14,7 @@ data class GutenbergPropsBuilder(
     private val enableUnsupportedBlockEditor: Boolean,
     private val unsupportedBlockEditorSwitch: Boolean,
-    private val enableAudioBlock: Boolean,
+    private val isAudioBlockMediaUploadEnabled: Boolean,
```
These changes come from this [PR](https://github.com/wordpress-mobile/WordPress-Android/pull/14680), that it was required to merge it here due to conflicts in the Gutenberg reference.

fluiddot commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergPropsBuilder.kt:
```diff
@@ -29,7 +29,7 @@ data class GutenbergPropsBuilder(
             enableUnsupportedBlockEditor = enableUnsupportedBlockEditor,
             canEnableUnsupportedBlockEditor = unsupportedBlockEditorSwitch,
-            enableAudioBlock = enableAudioBlock,
+            isAudioBlockMediaUploadEnabled = isAudioBlockMediaUploadEnabled,
```
These changes come from this [PR](https://github.com/wordpress-mobile/WordPress-Android/pull/14680), that it was required to merge it here due to conflicts in the Gutenberg reference.

SiobhyB commented:
To close the loop: I've created an issue for the Android Studio error here: https://github.com/wordpress-mobile/WordPress-Android/issues/14773. As I'll be working on that class as part of my HACK week project, I can spend some time trying to see if it's something I can figure out next week.
